mkosi/__init__.py

```diff
@@ -2144,66 +2148,82 @@ def maybe_compress(
-        return
+        return None
+
+    uki = context.root / efi_boot_binary(context)
+    if uki.exists() and (KernelType.identify(context.config, uki) == KernelType.uki):
+        return uki
+
+    uki = context.root / shim_second_stage_binary(context)
+    if uki.exists() and (KernelType.identify(context.config, uki) == KernelType.uki):
+        return uki
 
     ukis = sorted(
         (context.root / "boot/EFI/Linux").glob("*.efi"),
         key=lambda p: GenericVersion(p.name),
         reverse=True,
     )
 
-    if (uki := context.root / efi_boot_binary(context)).exists() and (
-        KernelType.identify(context.config, uki) == KernelType.uki
-    ):
-        pass
-    elif (uki := context.root / shim_second_stage_binary(context)).exists() and (
-        KernelType.identify(context.config, uki) == KernelType.uki
-    ):
-        pass
-    elif ukis:
-        uki = ukis[0]
-    else:
-        return
+    if ukis:
+        return ukis[0]
 
-    shutil.copy(uki, context.staging / context.config.output_split_uki)
+    return None
 
-    # Extract the combined initrds from the UKI so we can use it to direct kernel boot with qemu if needed.
-    extract_pe_section(context, uki, ".initrd", context.staging / context.config.output_split_initrd)
 
-    # ukify will have signed the kernel image as well. Let's make sure we put the signed kernel
-    # image in the output directory instead of the unsigned one by reading it from the UKI.
-    extract_pe_section(context, uki, ".linux", context.staging / context.config.output_split_kernel)
+def copy_uki(context: Context) -> None:
+    if ArtifactOutput.uki not in context.config.split_artifacts:
+        return
+
+    if (context.staging / context.config.output_split_uki).exists():
+        return
+
+    if uki := get_uki(context):
+        shutil.copy(uki, context.staging / context.config.output_split_uki)
 
 
 def copy_vmlinuz(context: Context) -> None:
+    if ArtifactOutput.kernel not in context.config.split_artifacts:
+        return
+
     if (context.staging / context.config.output_split_kernel).exists():
         return
 
+    # ukify will have signed the kernel image as well. Let's make sure we put the signed kernel
+    # image in the output directory instead of the unsigned one by reading it from the UKI.
+    if uki := get_uki(context):
+        extract_pe_section(context, uki, ".linux", context.staging / context.config.output_split_kernel)
+        return
+
     for _, kimg in gen_kernel_images(context):
         shutil.copy(context.root / kimg, context.staging / context.config.output_split_kernel)
         break
 
 
-def copy_nspawn_settings(context: Context) -> None:
-    if context.config.nspawn_settings is None:
-        return None
-
-    with complete_step("Copying nspawn settings file…"):
-        shutil.copy2(context.config.nspawn_settings, context.staging / context.config.output_nspawn_settings)
-
-
```

Comment:
Nit: it makes sense to move the function from this somewhat random place, having that as the first commit would be nicer, since that obscures the refactor a bit and doesn't have any functional change.